python/lsst/meas/extensions/scarlet/io.py

```diff
@@ -547,6 +548,16 @@ def _model_to_zip(self, model_data: scl.io.ScarletModelData) -> dict[str, Any]:
         })
         return data
 
+    def read_from_local_file(self, path: str, component: str | None = None, expected_size: int = -1) -> Any:
+        # Override of `FormatterV2.read_from_local_file`.
+        with zipfile.ZipFile(path, 'r') as zip_file:
+            unzipped_files = {}
```

Comment:
I don't like the code duplication here.  Why isn't there a `ScarletModelData.from_stream(stream, blend_ids)` API which is called from `ScarletModelData.from_file(path, blend_ids)` that the formatter calls? How does someone read a file they are sent that is not in butler?

Reply:
Scarlet lite models are typically stored as a single JSON object, so `ScarletModelData` has a `parse_obj` method that extracts all of the model data from JSON into the model data objects. I implemented the zip storage in `meas_extensions_scarlet` and not `scarlet_lite` because I don't want to tie scarlet lite users to any file storage system, so this implementation is only for the science pipelines.

Reply:
I am concerned that the same data is stored in two different formats but, importantly, if someone passes a file in this .scarlet zip format they can't read it using pipelines code. What's your strategy for how you tell people how to read a file? We can't stop people passing files around outside of butler.

I still don't like the code duplication here though. Please delegate the local file read to the stream reading (and you can factor the stream reading out into a class method if you are not adding a method to ScarletModelData).

Reply:
I do feel strongly about this because it's core logic on how to extract the model from the file. Private methods don't need as much documentation but in this case I don't see a problem with a factored out `_read_stream()` private method, with five lines of docstring, that is called directly from the `read_from_stream()` and via `with open() as fd:` call in `read_from_local_file`.

Reply:
Ok, what about this: I can move `_build_model`, `_model_to_zip` from class methods to public functions and also create a `read_scarlet_model` method that reads the file. Then the formatter can call that method from `read_from_local_file` and `read_from_stream`. I can also create a `write_scarlet_model` function that writes to a file and just call that from `ScarletModelFormatter.to_bytes`. Is that ok?

Reply:
That would work. Still somewhat confused that these aren't methods on `ScarletModelData` itself but public free functions are better than having to tell people to instantiate a formatter to read a file standalone.

Reply:
The main reason I don't include them in scarlet lite is because (as we saw in RFC-1074) there are numerous ways to save hierarchical data, including packages to save zip files. For example, HDF required a different pre-formatting before writing and reading than zip, and ASDF had a different pre-formatting. Given infinite time I'd just implement them all but since we're only implementing zip in the science pipelines it makes more sense to me to keep it in `meas_extensions_scarlet` as opposed to `scarlet_lite`.